### ISSUE TITLE
Exporting TypedArray vs Array<number>

### ISSUE THREAD

**Issue by jtenner** [AssemblyScript/assemblyscript#339]
```ts
export const TileMap = new Uint8Array(100 * 75);
export const SpriteMap = new Float64Array(100 * 7);
export const Particles = new Array<f64>(100 * 3);
```

Instead of allocating a ton of memory every frame for a wasm game I'm building, I decided to try and export an array from the module. The `Particles` array here works perfectly as expected, returning a pointer to an array definition, but for some reason when I check `exports.SpriteMap.value` and `exports.TileMap.value`, they both return `0`.

Also, If I remove the `export` keyword on any of the TypedArrays, and manually return a pointer to it from within a function, it works as intended. The export keyword will set the pointer to 0.

Is there a reason for this behavior, or am I missing something?

**Comment by dcodeIO** [comment]
I think the issue you are seeing there comes from WebAssembly not yet supporting mutable globals, and the AssemblyScript compiler not making a static memory segment from a `new` expression (yet) that would have an actual constant value. Using a function to return the reference is the proper way to do it right now, though I wonder why this didn't result in a compiler warning - or did it?

**Comment by jtenner** [comment]
It resulted in a compiler warning, but let's take the following example:

```ts
 export const Particles = new Array<f64>(1000 * 3);
```

This yields a warning, and yet it returns the proper pointer at runtime.

```
WARNING AS206: Compiling constant with non-constant initializer as mutable.

 export const Particles = new Array<f64>(1000 * 3);
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 in assembly/index.ts(7,13)
```

**Comment by MaxGraey** [comment]
Actually not all browsers support mutable globals. AS could enable this post MVP feature via `--feature mutable-global` but I don't sure this properly works.

That's pretty unexpected approach btw. Usually (and may this best approach) static allocation happened on JS side, and sending offsets (pointers) to wasm. See how this happening on [this example](https://github.com/AssemblyScript/assemblyscript/blob/master/examples/game-of-life/index.html#L81)

**Comment by jtenner** [comment]
@MaxGraey I was simply noting that this approach works with wasm allocated `Array<type>`, and it appears inconsistent. As for unexpected approach, is there a reason why this is the way it is?

**Comment by MaxGraey** [comment]
You approach required this proposal: https://github.com/WebAssembly/mutable-global. Which is only recently standardized.

**Comment by jtenner** [comment]
I see. I guess this doesn't answer my question about why `Float64Array` returns 0 when exported, vs when `Array<f64>` returns a pointer. I wasn't sure if this was undefined behavior, or a bug, or my fault.

I'm happy to return pointers from functions.

Thanks for your time!

**Comment by MaxGraey** [comment]
`Array<f64>` could works because it allocated first and reference to zero as well as other mutable globals, so it just a coincidence. If you change `HEAP_BASE` via `--memoryBase` flag to something different than `0` I guess `Array<f64>` also shouldn't work

**Comment by jtenner** [comment]
Ah! So it is somewhere between undefined behavior/bug. I made the switch to the following snippet, and the compiler seems to like it.

```ts
let TileMap = new Array<u8>(100 * 75);
export function getTileMap(): u8[] { return TileMap; }
```

I just wanted to point out something that I noticed relative to this problem. The reason why I chose `const` is because of tslint best practices.  If I use `let` instead of const in this pattern (which was suggested, right?), and treat the variable `TileMap` immutably, tslint swoops in and complains because it's never reassigned. I get the following message:

```
Identifier 'TileMap' is never reassigned; use 'const' instead of 'let'.
```

So the workaround is obviously to suppress tslint, with a comment surrounding the globals, but I was wondering if something else could be done?
